sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -331,7 +360,10 @@ def set_request(estimator, method_name):
             set_request(estimator, method_name)
             instance = cls(**kwargs)
             method = getattr(instance, method_name)
-            method(X, y, **method_kwargs)
+            try:
+                method(X, y, **method_kwargs)
+            except ValueError:
+                method(X, y, classes=np.unique(y), **method_kwargs)
```

Comment:
you could reverse the order here, and catch `TypeError` and always pass `classes` if it's in the signature.

Reply:
@adrinjalali Thanks for the suggestion. Could you please explain the advantages of reversing the order? I don't see it, unfortunately.

My thought was to try what covers the most cases first to potentially save computational resources. Vastly most meta-estimators don't have `classes` implemented.

Reply:
we're basically checking if the signature of the function supports `classes` or not, if yes, we pass it. If the signature doesn't support it, we get a `TypeError` which is standard in Python. Relying on the `ValueError` is something in our code instead of standard python machinery.

Reply:
> Relying on the ValueError is something in our code instead of standard python machinery.

I'm sorry, but I can't follow your reasoning and I'd like to learn and understand what you ask me to change. Why is turning it around to check the least common case first, better? Why is having TypeError more desirable than ValueError?

Reply:
One is raised by Python machinery, one is raised by our code, and other estimator might raise a different error than `ValueError`. Here relying on python's machinery is much more resilient and reliable than relying on code that we write in our estimators.

Reply:
This now makes sense to me. :bulb:  Thank you. I will change it with the next push later today.